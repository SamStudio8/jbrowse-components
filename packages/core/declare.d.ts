declare module 'shortid'
declare module 'object.fromentries'
