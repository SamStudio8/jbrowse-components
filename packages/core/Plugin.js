export default class Plugin {
  install() {
    this.installed = true
  }

  configure() {
    this.configured = true
  }
}
