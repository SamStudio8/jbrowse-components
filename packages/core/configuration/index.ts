export {
  ConfigurationSchema,
  ConfigurationSchemaType,
  ConfigurationSchemaDefinition,
  ConfigurationModel,
  ConfigurationReference,
  isConfigurationModel,
  isConfigurationSchemaType,
  isBareConfigurationSchemaType,
} from './configurationSchema'

export { getConf, readConfObject } from './util'
