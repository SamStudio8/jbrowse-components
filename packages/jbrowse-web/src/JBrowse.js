import '@gmod/jbrowse-core/fonts/material-icons.css'
import CircularProgress from '@material-ui/core/CircularProgress'
import CssBaseline from '@material-ui/core/CssBaseline'
import { ThemeProvider } from '@material-ui/styles'
import { observer } from 'mobx-react'
import React from 'react'
import 'typeface-roboto'
import App from './ui/App'
import Theme from './ui/theme'

export default observer(({ state }) => {
  let DisplayComponent = (
    <CircularProgress
      style={{
        position: 'fixed',
        top: '50%',
        left: '50%',
        marginTop: -250,
        marginLeft: -250,
      }}
      size={500}
    />
  )
  if (state.errorMessage) DisplayComponent = <div>{state.errorMessage}</div>
  if (state.activeSession)
    DisplayComponent = (
      <App
        session={state.activeSession}
        sessionNames={state.sessionNames}
        setActiveSession={state.activateSession}
        addSession={state.addSession}
      />
    )

  return (
    <ThemeProvider theme={Theme}>
      <CssBaseline />
      {DisplayComponent}
    </ThemeProvider>
  )
})
