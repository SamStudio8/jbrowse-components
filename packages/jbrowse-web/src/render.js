import Rpc from '@librpc/web'

import SimpleFeature from './util/feature'

// import { isStateTreeNode, getSnapshot } from 'mobx-state-tree'

export function fog() {}

// // given an object of things that might be MST models, snapshots any of them
// // that are not already snapshots
// function ensureSnapshots(things) {
//   const data = {}
//   Object.entries(things).forEach(([key, value]) => {
//     if (isStateTreeNode(value)) {
//       data[key] = getSnapshot(value)
//     } else {
//       data[key] = value
//     }
//   })
//   return data
// }

const getClient = pluginManager =>
  new Rpc.Client({
    workers: pluginManager.getWorkerGroup('render'),
  })

export async function renderRegionWithWorker(pluginManager, args) {
  const result = await getClient(pluginManager).call('renderRegion', args, {
    timeout: args.timeout,
  })

  // convert the feature JSON to SimpleFeature objects
  result.features = result.featureJSON.map(j => SimpleFeature.fromJSON(j))
  return result
}

export async function freeSessionResourcesInWorker(pluginManager, sessionId) {
  return getClient(pluginManager).call('freeSessionResources', sessionId)
}
