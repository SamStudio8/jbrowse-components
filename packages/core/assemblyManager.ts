import AbortablePromiseCache from 'abortable-promise-cache'
import jsonStableStringify from 'json-stable-stringify'
import { autorun, observable } from 'mobx'
import {
  addDisposer,
  cast,
  getParent,
  IAnyType,
  SnapshotOrInstance,
  types,
} from 'mobx-state-tree'
import { readConfObject } from './configuration'
import { AnyConfigurationModel } from './configuration/configurationSchema'
import { Region } from './util/types'
import { Region as MSTRegion } from './util/types/mst'
import QuickLRU from './util/QuickLRU'

function checkRefName(refName: string) {
  // Valid refName pattern from https://samtools.github.io/hts-specs/SAMv1.pdf
  if (
    !refName.match(
      /[0-9A-Za-z!#$%&+./:;?@^_|~-][0-9A-Za-z!#$%&*+./:;=?@^_|~-]*/,
    )
  ) {
    throw new Error(`Encountered invalid refName: "${refName}"`)
  }
}

export function assemblyFactory(assemblyConfigType: IAnyType) {
  return types
    .model({
      configuration: types.reference(assemblyConfigType),
      regions: types.array(MSTRegion),
      refNameAliases: types.map(types.array(types.string)),
      refNameAliasesSet: false,
    })
    .views(self => ({
      get name(): string {
        return readConfObject(self.configuration, 'name')
      },
      get aliases(): string[] {
        return readConfObject(self.configuration, 'aliases')
      },
      get refNames() {
        return self.regions.map(region => region.refName)
      },
      get ready() {
        return !!self.regions.length && self.refNameAliasesSet
      },
      get allRefNames() {
        const aliases: string[] = []
        self.refNameAliases.forEach(aliasList => {
          aliases.push(...aliasList)
        })
        return [...this.refNames, ...aliases]
      },
      get rpcManager() {
        return getParent(self, 2).rpcManager
      },
      getCanonicalRefName(refName: string) {
        if (this.refNames.includes(refName)) {
          return refName
        }
        for (const [rName, aliases] of self.refNameAliases) {
          if (aliases.includes(refName)) {
            return rName
          }
        }
        return undefined
      },
      isValidRefName(refName: string) {
        return this.allRefNames.includes(refName)
      },
    }))
    .actions(self => ({
      afterAttach() {
        addDisposer(
          self,
          autorun(async () => {
            const sequenceAdapterConfig = readConfObject(self.configuration, [
              'sequence',
              'adapter',
            ])
            const adapterConfigId = jsonStableStringify(sequenceAdapterConfig)
            const sequenceAborter = new AbortController()
            const { rpcManager } = getParent(self, 2)
            rpcManager
              .call(
                adapterConfigId,
                'getRegions',
                {
                  sessionId: self.name,
                  adapterConfig: sequenceAdapterConfig,
                  signal: sequenceAborter.signal,
                },
                { timeout: 1000000 },
              )
              .then((adapterRegions: Region[]) => {
                const adapterRegionsWithAssembly = adapterRegions.map(
                  adapterRegion => {
                    const { refName } = adapterRegion
                    checkRefName(refName)
                    return {
                      ...adapterRegion,
                      assemblyName: self.name,
                    }
                  },
                )
                this.setRegions(adapterRegionsWithAssembly)
              })
              .catch((error: Error) => {
                getParent(self, 3).setError(String(error))
              })
            if (self.configuration.refNameAliases) {
              const refNameAliasesAdapterConfig = readConfObject(
                self.configuration,
                ['refNameAliases', 'adapter'],
              )
              const refNameAliasesAborter = new AbortController()
              rpcManager
                .call(
                  adapterConfigId,
                  'getRefNameAliases',
                  {
                    sessionId: self.name,
                    adapterConfig: refNameAliasesAdapterConfig,
                    signal: refNameAliasesAborter.signal,
                  },
                  { timeout: 1000000 },
                )
                .then(
                  (
                    refNameAliasesList: {
                      refName: string
                      aliases: string[]
                    }[],
                  ) => {
                    const refNameAliases: Record<string, string[]> = {}
                    refNameAliasesList.forEach(refNameAlias => {
                      refNameAlias.aliases.forEach(alias => {
                        checkRefName(alias)
                      })
                      refNameAliases[refNameAlias.refName] =
                        refNameAlias.aliases
                    })
                    this.setRefNameAliases(refNameAliases)
                    this.setRefNameAliasesSet(true)
                  },
                )
                .catch((error: unknown) => {
                  getParent(self, 3).setError(String(error))
                })
            } else {
              this.setRefNameAliasesSet(true)
            }
          }),
        )
      },
      setRegions(regions: Region[]) {
        self.regions = cast(regions)
      },
      setRefNameAliases(refNameAliases: Record<string, string[]>) {
        self.refNameAliases = cast(refNameAliases)
      },
      setRefNameAliasesSet(isSet: boolean) {
        self.refNameAliasesSet = isSet
      },
    }))
    .actions(self => {
      const refNameMapCache = new AbortablePromiseCache({
        // QuickLRU is a good backing cache to use, but you can use any
        // cache as long as it supports `get`, `set`, `delete`, and `keys`.
        cache: new QuickLRU({ maxSize: 1000 }),

        // the `fill` callback will be called for a cache miss
        async fill(
          { adapterConf }: { adapterConf: unknown },
          signal?: AbortSignal,
        ) {
          const refNameMap = observable.map({})
          if (!self.refNameAliasesSet) {
            return refNameMap
          }
          const stateGroupName = jsonStableStringify(adapterConf)

          const refNames = await self.rpcManager.call(
            stateGroupName,
            'getRefNames',
            {
              sessionId: self.name,
              adapterConfig: adapterConf,
              signal,
            },
            { timeout: 1000000 },
          )
          refNames.forEach((refName: string) => {
            checkRefName(refName)
            const aliases = self.refNameAliases.get(refName)
            if (aliases) {
              aliases.forEach(refNameAlias => {
                refNameMap.set(refNameAlias, refName)
              })
            } else {
              self.refNameAliases.forEach((configAliases, configRefName) => {
                if (configAliases.includes(refName)) {
                  refNameMap.set(configRefName, refName)
                  configAliases.forEach(refNameAlias => {
                    if (refNameAlias !== refName)
                      refNameMap.set(refNameAlias, refName)
                  })
                }
              })
            }
          })
          return refNameMap
        },
      })

      /**
       * get Map of `canonical-name -> adapter-specific-name`
       */
      async function getRefNameMapForAdapter(
        adapterConf: unknown,
        opts: { signal?: AbortSignal } = {},
      ) {
        const adapterConfigId = jsonStableStringify(adapterConf)
        return refNameMapCache.get(
          adapterConfigId,
          { adapterConf },
          opts.signal,
        )
      }

      /**
       * get Map of `adapter-specific-name -> canonical-name`
       */
      async function getReverseRefNameMapForAdapter(
        adapterConf: unknown,
        opts: { signal?: AbortSignal } = {},
      ) {
        const map = await getRefNameMapForAdapter(adapterConf, opts)
        const reversed = new Map()
        for (const [canonicalName, adapterName] of map) {
          reversed.set(adapterName, canonicalName)
        }
        return reversed
      }
      return { getRefNameMapForAdapter, getReverseRefNameMapForAdapter }
    })
}

export default function assemblyManagerFactory(assemblyConfigType: IAnyType) {
  return types
    .model({
      assemblies: types.array(assemblyFactory(assemblyConfigType)),
    })
    .views(self => ({
      get(assemblyName: string) {
        const canonicalName = this.aliasMap.get(assemblyName)
        return self.assemblies.find(
          assembly => assembly.name === (canonicalName || assemblyName),
        )
      },
      get aliasMap() {
        const aliases: Map<string, string> = new Map()
        self.assemblies.forEach(assembly => {
          if (assembly.aliases.length) {
            assembly.aliases.forEach(assemblyAlias => {
              aliases.set(assemblyAlias, assembly.name)
            })
          }
        })
        return aliases
      },
      get rpcManager() {
        return getParent(self).rpcManager
      },
      get allPossibleRefNames() {
        const refNames: string[] = []
        self.assemblies.forEach(assembly => {
          refNames.push(...assembly.refNames)
        })
        return refNames
      },
    }))
    .views(self => ({
      async getRefNameMapForAdapter(
        adapterConf: unknown,
        assemblyName: string,
        opts: { signal?: AbortSignal } = {},
      ) {
        const assembly = self.get(assemblyName)
        if (assembly) {
          return assembly.getRefNameMapForAdapter(adapterConf, opts)
        }
        return observable.map({})
      },
      async getReverseRefNameMapForAdapter(
        adapterConf: unknown,
        assemblyName: string,
        opts: { signal?: AbortSignal } = {},
      ) {
        const assembly = self.get(assemblyName)
        if (assembly) {
          return assembly.getReverseRefNameMapForAdapter(adapterConf, opts)
        }
        return observable.map({})
      },
      isValidRefName(refName: string, assemblyName?: string) {
        if (assemblyName) {
          const assembly = self.get(assemblyName)
          if (assembly) {
            return assembly.isValidRefName(refName)
          }
        }
        return self.allPossibleRefNames.includes(refName)
      },
    }))
    .actions(self => ({
      afterAttach() {
        // self.assemblies.push({
        //   configuration: getParent(self).jbrowse.assemblies[0],
        // })
        addDisposer(
          self,
          autorun(() => {
            getParent(self).jbrowse.assemblies.forEach(
              (assemblyConfig: AnyConfigurationModel) => {
                const existingAssemblyIdx = self.assemblies.findIndex(
                  assembly =>
                    assembly.name === readConfObject(assemblyConfig, 'name'),
                )
                if (existingAssemblyIdx !== -1) {
                  this.replaceAssembly(existingAssemblyIdx, assemblyConfig)
                } else {
                  this.addAssembly(assemblyConfig)
                }
              },
            )
          }),
        )
      },
      addAssembly(
        assemblyConfig: SnapshotOrInstance<typeof assemblyConfigType> | string,
      ) {
        self.assemblies.push({ configuration: assemblyConfig })
      },
      replaceAssembly(
        idx: number,
        assemblyConfig: SnapshotOrInstance<typeof assemblyConfigType> | string,
      ) {
        self.assemblies[idx] = assemblyConfigType.create({
          configuration: assemblyConfig,
        })
      },
    }))
}
