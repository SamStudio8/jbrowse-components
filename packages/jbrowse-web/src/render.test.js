import { ExpansionPanelActions } from '@material-ui/core'
import JBrowse from './JBrowse'
import { renderRegion } from './render'

const jbrowse = new JBrowse().configure()

test('can render a single region with Pileup + BamAdapter', async () => {
  const testprops = {
    region: { assembly: 'volvox', refName: 'ctgA', start: 0, end: 800 },
    adapterType: 'BamAdapter',
    adapterConfig: {
      _configId: '7Hc9NkuD4x',
      type: 'BamAdapter',
      bamLocation: {
        path: require.resolve('../public/test_data/volvox-sorted.bam'),
      },
      index: {
        _configId: 'sGW8va26pr',
        location: {
          path: require.resolve('../public/test_data/volvox-sorted.bam.bai'),
        },
      },
    },
    rendererType: 'PileupRenderer',
    renderProps: {},
  }

  const result = await renderRegion(jbrowse, testprops)
  expect(Object.keys(result)).toEqual(['features', 'html'])
  expect(result.features.length).toBe(93)
  expect(result.html).toMatchSnapshot()
})
