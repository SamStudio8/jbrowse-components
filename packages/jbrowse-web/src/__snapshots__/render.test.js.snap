// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can render a single region with Pileup + BamAdapter 1`] = `"<div class=\\"PileupRendering\\" style=\\"position:relative\\" data-reactroot=\\"\\"><canvas width=\\"800\\" height=\\"144\\"></canvas><canvas width=\\"800\\" height=\\"144\\" style=\\"position:absolute;left:0;top:0\\" class=\\"highlightOverlayCanvas\\"></canvas></div>"`;

exports[`can render a single region with Pileup + BamAdapter 2`] = `
Object {
  "rectangles": Object {
    "-1055583960": Array [
      630,
      138,
      730,
      143,
    ],
    "-1150282940": Array [
      413,
      18,
      513,
      23,
    ],
    "-1157339540": Array [
      764,
      0,
      864,
      5,
    ],
    "-1199599379": Array [
      713,
      6,
      813,
      11,
    ],
    "-1223356361": Array [
      797,
      36,
      897,
      41,
    ],
    "-1240580270": Array [
      109,
      0,
      209,
      5,
    ],
    "-126415083": Array [
      123,
      6,
      223,
      11,
    ],
    "-1280931855": Array [
      570,
      54,
      670,
      59,
    ],
    "-1294921403": Array [
      574,
      96,
      674,
      101,
    ],
    "-1327260758": Array [
      42,
      30,
      142,
      35,
    ],
    "-1402357424": Array [
      90,
      66,
      190,
      71,
    ],
    "-1543509453": Array [
      422,
      30,
      522,
      35,
    ],
    "-1573216292": Array [
      47,
      42,
      147,
      47,
    ],
    "-1578361879": Array [
      625,
      120,
      725,
      125,
    ],
    "-1590347580": Array [
      302,
      24,
      402,
      29,
    ],
    "-1676418289": Array [
      582,
      60,
      682,
      65,
    ],
    "-1682306516": Array [
      513,
      78,
      613,
      83,
    ],
    "-1706997693": Array [
      690,
      48,
      790,
      53,
    ],
    "-1730672711": Array [
      626,
      126,
      726,
      131,
    ],
    "-1861902917": Array [
      661,
      24,
      761,
      29,
    ],
    "-1903522961": Array [
      461,
      48,
      561,
      53,
    ],
    "-233455955": Array [
      611,
      12,
      711,
      17,
    ],
    "-262734143": Array [
      775,
      24,
      875,
      29,
    ],
    "-397495927": Array [
      577,
      102,
      677,
      107,
    ],
    "-432814408": Array [
      17,
      24,
      117,
      29,
    ],
    "-467929454": Array [
      453,
      0,
      553,
      5,
    ],
    "-487480200": Array [
      7,
      6,
      107,
      11,
    ],
    "-517500600": Array [
      725,
      60,
      825,
      65,
    ],
    "-526084999": Array [
      76,
      60,
      176,
      65,
    ],
    "-541132950": Array [
      664,
      30,
      764,
      35,
    ],
    "-54442761": Array [
      509,
      12,
      609,
      17,
    ],
    "-584714253": Array [
      413,
      24,
      513,
      29,
    ],
    "-624523461": Array [
      152,
      42,
      252,
      47,
    ],
    "-632885881": Array [
      354,
      54,
      454,
      59,
    ],
    "-724923975": Array [
      562,
      30,
      662,
      35,
    ],
    "-783205836": Array [
      564,
      42,
      664,
      47,
    ],
    "-866123030": Array [
      729,
      72,
      829,
      77,
    ],
    "-924959209": Array [
      125,
      12,
      225,
      17,
    ],
    "-94743417": Array [
      786,
      30,
      886,
      35,
    ],
    "-970365713": Array [
      258,
      6,
      358,
      11,
    ],
    "-999079398": Array [
      321,
      36,
      421,
      41,
    ],
    "1004747877": Array [
      556,
      0,
      656,
      5,
    ],
    "1046677334": Array [
      730,
      78,
      830,
      83,
    ],
    "1079933533": Array [
      766,
      18,
      866,
      23,
    ],
    "1086146284": Array [
      685,
      42,
      785,
      47,
    ],
    "1114949599": Array [
      149,
      36,
      249,
      41,
    ],
    "1150161791": Array [
      16,
      18,
      116,
      23,
    ],
    "1213790853": Array [
      132,
      18,
      232,
      23,
    ],
    "1272673179": Array [
      595,
      72,
      695,
      77,
    ],
    "1358701805": Array [
      399,
      12,
      499,
      17,
    ],
    "1390838279": Array [
      15,
      12,
      115,
      17,
    ],
    "1401207929": Array [
      660,
      0,
      760,
      5,
    ],
    "1410870858": Array [
      727,
      66,
      827,
      71,
    ],
    "1449206326": Array [
      467,
      54,
      567,
      59,
    ],
    "1478804620": Array [
      291,
      12,
      391,
      17,
    ],
    "1480692980": Array [
      140,
      24,
      240,
      29,
    ],
    "1514689949": Array [
      579,
      108,
      679,
      113,
    ],
    "1542598533": Array [
      55,
      48,
      155,
      53,
    ],
    "1559390564": Array [
      679,
      36,
      779,
      41,
    ],
    "1615231956": Array [
      311,
      30,
      411,
      35,
    ],
    "1629308221": Array [
      720,
      54,
      820,
      59,
    ],
    "1697419078": Array [
      432,
      42,
      532,
      47,
    ],
    "1806471452": Array [
      147,
      30,
      247,
      35,
    ],
    "1846631799": Array [
      593,
      66,
      693,
      71,
    ],
    "1882240796": Array [
      611,
      6,
      711,
      11,
    ],
    "208318216": Array [
      65,
      54,
      165,
      59,
    ],
    "2112054817": Array [
      564,
      48,
      664,
      53,
    ],
    "212318835": Array [
      325,
      42,
      425,
      47,
    ],
    "2135426977": Array [
      2,
      0,
      102,
      5,
    ],
    "21405281": Array [
      327,
      48,
      427,
      53,
    ],
    "230559949": Array [
      390,
      6,
      490,
      11,
    ],
    "230630259": Array [
      94,
      72,
      194,
      77,
    ],
    "298547082": Array [
      338,
      0,
      438,
      5,
    ],
    "374397043": Array [
      652,
      18,
      752,
      23,
    ],
    "431313551": Array [
      566,
      84,
      666,
      89,
    ],
    "49754882": Array [
      431,
      36,
      531,
      41,
    ],
    "529426168": Array [
      612,
      114,
      712,
      119,
    ],
    "53128272": Array [
      715,
      12,
      815,
      17,
    ],
    "531931475": Array [
      545,
      24,
      645,
      29,
    ],
    "543331399": Array [
      237,
      0,
      337,
      5,
    ],
    "640474988": Array [
      564,
      36,
      664,
      41,
    ],
    "679424769": Array [
      574,
      90,
      674,
      95,
    ],
    "687062757": Array [
      630,
      132,
      730,
      137,
    ],
    "68768545": Array [
      618,
      78,
      718,
      83,
    ],
    "710214930": Array [
      301,
      18,
      401,
      23,
    ],
    "748240882": Array [
      481,
      60,
      581,
      65,
    ],
    "750545343": Array [
      747,
      84,
      847,
      89,
    ],
    "881925482": Array [
      540,
      18,
      640,
      23,
    ],
    "904949709": Array [
      43,
      36,
      143,
      41,
    ],
    "911993950": Array [
      487,
      72,
      587,
      77,
    ],
    "93246257": Array [
      485,
      66,
      585,
      71,
    ],
    "93942220": Array [
      495,
      6,
      595,
      11,
    ],
    "941345053": Array [
      186,
      48,
      286,
      53,
    ],
  },
  "totalHeight": 144,
}
`;

exports[`can render a single region with SvgFeatures + BamAdapter 1`] = `"<svg class=\\"SvgFeatureRendering\\" width=\\"300px\\" height=\\"648px\\" style=\\"position:relative\\" data-reactroot=\\"\\"><g transform=\\"translate(2 0)\\"><rect title=\\"2135426977\\" x=\\"0\\" y=\\"0\\" width=\\"100\\" height=\\"10\\" style=\\"fill:goldenrod\\"></rect><text x=\\"0\\" y=\\"12\\" style=\\"font-size:13px;fill:black\\" dominant-baseline=\\"hanging\\">ctgA_3_555_0:0:0_2:0:0_102d</text></g><g transform=\\"translate(7 27)\\"><rect title=\\"-487480200\\" x=\\"0\\" y=\\"0\\" width=\\"100\\" height=\\"10\\" style=\\"fill:goldenrod\\"></rect><text x=\\"0\\" y=\\"12\\" style=\\"font-size:13px;fill:black\\" dominant-baseline=\\"hanging\\">ctgA_8_507_4:0:0_2:0:0_10e6</text></g><g transform=\\"translate(15 54)\\"><rect title=\\"1390838279\\" x=\\"0\\" y=\\"0\\" width=\\"100\\" height=\\"10\\" style=\\"fill:goldenrod\\"></rect><text x=\\"0\\" y=\\"12\\" style=\\"font-size:13px;fill:black\\" dominant-baseline=\\"hanging\\">ctgA_16_523_3:0:0_0:0:0_1a75</text></g><g transform=\\"translate(16 81)\\"><rect title=\\"1150161791\\" x=\\"0\\" y=\\"0\\" width=\\"100\\" height=\\"10\\" style=\\"fill:goldenrod\\"></rect><text x=\\"0\\" y=\\"12\\" style=\\"font-size:13px;fill:black\\" dominant-baseline=\\"hanging\\">ctgA_17_531_4:0:0_2:0:0_b65</text></g><g transform=\\"translate(17 108)\\"><rect title=\\"-432814408\\" x=\\"0\\" y=\\"0\\" width=\\"100\\" height=\\"10\\" style=\\"fill:goldenrod\\"></rect><text x=\\"0\\" y=\\"12\\" style=\\"font-size:13px;fill:black\\" dominant-baseline=\\"hanging\\">ctgA_18_445_3:0:0_2:0:0_22df</text></g><g transform=\\"translate(42 135)\\"><rect title=\\"-1327260758\\" x=\\"0\\" y=\\"0\\" width=\\"100\\" height=\\"10\\" style=\\"fill:goldenrod\\"></rect><text x=\\"0\\" y=\\"12\\" style=\\"font-size:13px;fill:black\\" dominant-baseline=\\"hanging\\">ctgA_43_557_1:0:0_2:0:0_5b3</text></g><g transform=\\"translate(43 162)\\"><rect title=\\"904949709\\" x=\\"0\\" y=\\"0\\" width=\\"100\\" height=\\"10\\" style=\\"fill:goldenrod\\"></rect><text x=\\"0\\" y=\\"12\\" style=\\"font-size:13px;fill:black\\" dominant-baseline=\\"hanging\\">ctgA_44_560_2:0:0_5:0:0_2524</text></g><g transform=\\"translate(47 189)\\"><rect title=\\"-1573216292\\" x=\\"0\\" y=\\"0\\" width=\\"100\\" height=\\"10\\" style=\\"fill:goldenrod\\"></rect><text x=\\"0\\" y=\\"12\\" style=\\"font-size:13px;fill:black\\" dominant-baseline=\\"hanging\\">ctgA_48_502_2:0:0_1:0:0_1513</text></g><g transform=\\"translate(55 216)\\"><rect title=\\"1542598533\\" x=\\"0\\" y=\\"0\\" width=\\"100\\" height=\\"10\\" style=\\"fill:goldenrod\\"></rect><text x=\\"0\\" y=\\"12\\" style=\\"font-size:13px;fill:black\\" dominant-baseline=\\"hanging\\">ctgA_56_576_1:0:0_2:0:0_191d</text></g><g transform=\\"translate(65 243)\\"><rect title=\\"208318216\\" x=\\"0\\" y=\\"0\\" width=\\"100\\" height=\\"10\\" style=\\"fill:goldenrod\\"></rect><text x=\\"0\\" y=\\"12\\" style=\\"font-size:13px;fill:black\\" dominant-baseline=\\"hanging\\">ctgA_66_630_1:0:0_3:0:0_ce5</text></g><g transform=\\"translate(76 270)\\"><rect title=\\"-526084999\\" x=\\"0\\" y=\\"0\\" width=\\"100\\" height=\\"10\\" style=\\"fill:goldenrod\\"></rect><text x=\\"0\\" y=\\"12\\" style=\\"font-size:13px;fill:black\\" dominant-baseline=\\"hanging\\">ctgA_77_676_0:0:0_3:0:0_764</text></g><g transform=\\"translate(90 297)\\"><rect title=\\"-1402357424\\" x=\\"0\\" y=\\"0\\" width=\\"100\\" height=\\"10\\" style=\\"fill:goldenrod\\"></rect><text x=\\"0\\" y=\\"12\\" style=\\"font-size:13px;fill:black\\" dominant-baseline=\\"hanging\\">ctgA_91_656_0:0:0_2:0:0_8f0</text></g><g transform=\\"translate(94 324)\\"><rect title=\\"230630259\\" x=\\"0\\" y=\\"0\\" width=\\"100\\" height=\\"10\\" style=\\"fill:goldenrod\\"></rect><text x=\\"0\\" y=\\"12\\" style=\\"font-size:13px;fill:black\\" dominant-baseline=\\"hanging\\">ctgA_95_545_0:0:0_1:0:0_1ecb</text></g><g transform=\\"translate(109 351)\\"><rect title=\\"-1240580270\\" x=\\"0\\" y=\\"0\\" width=\\"100\\" height=\\"10\\" style=\\"fill:goldenrod\\"></rect><text x=\\"0\\" y=\\"12\\" style=\\"font-size:13px;fill:black\\" dominant-baseline=\\"hanging\\">ctgA_110_638_0:0:0_3:0:0_15b</text></g><g transform=\\"translate(123 378)\\"><rect title=\\"-126415083\\" x=\\"0\\" y=\\"0\\" width=\\"100\\" height=\\"10\\" style=\\"fill:goldenrod\\"></rect><text x=\\"0\\" y=\\"12\\" style=\\"font-size:13px;fill:black\\" dominant-baseline=\\"hanging\\">ctgA_124_679_2:0:0_0:0:0_17d4</text></g><g transform=\\"translate(125 405)\\"><rect title=\\"-924959209\\" x=\\"0\\" y=\\"0\\" width=\\"100\\" height=\\"10\\" style=\\"fill:goldenrod\\"></rect><text x=\\"0\\" y=\\"12\\" style=\\"font-size:13px;fill:black\\" dominant-baseline=\\"hanging\\">ctgA_126_596_1:0:0_1:0:0_641</text></g><g transform=\\"translate(132 432)\\"><rect title=\\"1213790853\\" x=\\"0\\" y=\\"0\\" width=\\"100\\" height=\\"10\\" style=\\"fill:goldenrod\\"></rect><text x=\\"0\\" y=\\"12\\" style=\\"font-size:13px;fill:black\\" dominant-baseline=\\"hanging\\">ctgA_133_707_5:0:0_4:0:0_1062</text></g><g transform=\\"translate(140 459)\\"><rect title=\\"1480692980\\" x=\\"0\\" y=\\"0\\" width=\\"100\\" height=\\"10\\" style=\\"fill:goldenrod\\"></rect><text x=\\"0\\" y=\\"12\\" style=\\"font-size:13px;fill:black\\" dominant-baseline=\\"hanging\\">ctgA_141_644_1:0:0_0:0:0_2461</text></g><g transform=\\"translate(147 486)\\"><rect title=\\"1806471452\\" x=\\"0\\" y=\\"0\\" width=\\"100\\" height=\\"10\\" style=\\"fill:goldenrod\\"></rect><text x=\\"0\\" y=\\"12\\" style=\\"font-size:13px;fill:black\\" dominant-baseline=\\"hanging\\">ctgA_148_699_1:0:0_2:0:0_64d</text></g><g transform=\\"translate(149 513)\\"><rect title=\\"1114949599\\" x=\\"0\\" y=\\"0\\" width=\\"100\\" height=\\"10\\" style=\\"fill:goldenrod\\"></rect><text x=\\"0\\" y=\\"12\\" style=\\"font-size:13px;fill:black\\" dominant-baseline=\\"hanging\\">ctgA_150_624_3:0:0_3:0:0_2e</text></g><g transform=\\"translate(152 540)\\"><rect title=\\"-624523461\\" x=\\"0\\" y=\\"0\\" width=\\"100\\" height=\\"10\\" style=\\"fill:goldenrod\\"></rect><text x=\\"0\\" y=\\"12\\" style=\\"font-size:13px;fill:black\\" dominant-baseline=\\"hanging\\">ctgA_153_667_0:0:0_6:0:0_1f51</text></g><g transform=\\"translate(186 567)\\"><rect title=\\"941345053\\" x=\\"0\\" y=\\"0\\" width=\\"100\\" height=\\"10\\" style=\\"fill:goldenrod\\"></rect><text x=\\"0\\" y=\\"12\\" style=\\"font-size:13px;fill:black\\" dominant-baseline=\\"hanging\\">ctgA_187_683_2:1:0_7:0:0_9e1</text></g><g transform=\\"translate(237 594)\\"><rect title=\\"543331399\\" x=\\"0\\" y=\\"0\\" width=\\"100\\" height=\\"10\\" style=\\"fill:goldenrod\\"></rect><text x=\\"0\\" y=\\"12\\" style=\\"font-size:13px;fill:black\\" dominant-baseline=\\"hanging\\">ctgA_238_789_0:1:0_1:0:0_e0d</text></g><g transform=\\"translate(258 621)\\"><rect title=\\"-970365713\\" x=\\"0\\" y=\\"0\\" width=\\"100\\" height=\\"10\\" style=\\"fill:goldenrod\\"></rect><text x=\\"0\\" y=\\"12\\" style=\\"font-size:13px;fill:black\\" dominant-baseline=\\"hanging\\">ctgA_259_775_1:1:0_2:0:0_5fc</text></g><g transform=\\"translate(291 0)\\"><rect title=\\"1478804620\\" x=\\"0\\" y=\\"0\\" width=\\"100\\" height=\\"10\\" style=\\"fill:goldenrod\\"></rect><text x=\\"0\\" y=\\"12\\" style=\\"font-size:13px;fill:black\\" dominant-baseline=\\"hanging\\">ctgA_292_745_1:0:0_2:0:0_17f9</text></g></svg>"`;

exports[`can render a single region with SvgFeatures + BamAdapter 2`] = `
Object {
  "rectangles": Object {
    "-1240580270": Array [
      109,
      351,
      382,
      376,
    ],
    "-126415083": Array [
      123,
      378,
      405,
      403,
    ],
    "-1327260758": Array [
      42,
      135,
      305,
      160,
    ],
    "-1402357424": Array [
      90,
      297,
      353,
      322,
    ],
    "-1573216292": Array [
      47,
      189,
      320,
      214,
    ],
    "-432814408": Array [
      17,
      108,
      290,
      133,
    ],
    "-487480200": Array [
      7,
      27,
      270,
      52,
    ],
    "-526084999": Array [
      76,
      270,
      339,
      295,
    ],
    "-624523461": Array [
      152,
      540,
      434,
      565,
    ],
    "-924959209": Array [
      125,
      405,
      398,
      430,
    ],
    "-970365713": Array [
      258,
      621,
      531,
      646,
    ],
    "1114949599": Array [
      149,
      513,
      412,
      538,
    ],
    "1150161791": Array [
      16,
      81,
      279,
      106,
    ],
    "1213790853": Array [
      132,
      432,
      414,
      457,
    ],
    "1390838279": Array [
      15,
      54,
      288,
      79,
    ],
    "1478804620": Array [
      291,
      0,
      573,
      25,
    ],
    "1480692980": Array [
      140,
      459,
      422,
      484,
    ],
    "1542598533": Array [
      55,
      216,
      328,
      241,
    ],
    "1806471452": Array [
      147,
      486,
      420,
      511,
    ],
    "208318216": Array [
      65,
      243,
      328,
      268,
    ],
    "2135426977": Array [
      2,
      0,
      265,
      25,
    ],
    "230630259": Array [
      94,
      324,
      367,
      349,
    ],
    "543331399": Array [
      237,
      594,
      510,
      619,
    ],
    "904949709": Array [
      43,
      162,
      316,
      187,
    ],
    "941345053": Array [
      186,
      567,
      459,
      592,
    ],
  },
  "totalHeight": 648,
}
`;
